### ISSUE TITLE
s3 bucket to bucket copy operation failed, 1 GB files - Linking did not work

### ISSUE THREAD

**Issue by akmithal** [noobaa/noobaa-core#6565]
### Environment info
- NooBaa Version: master-20210607
- Platform: OCP 4.6.16

### Actual behavior
1. Copy operation from one bucket to another failed

### Expected behavior
1. Copy operation should not fail

### Steps to reproduce
1. Created 2 buckets - bucket-1, bucket-21.
2. Uploaded multiple 1GB files to s3://bucket-1
3. Attempted copy operation from s3://bucket-1 to s3://bucket-21

Few files got copied but got failure for others:

```
[root@hpo-node1 akshat]# s3u1 ls s3://bucket-1/
urllib3/connectionpool.py:1013: InsecureRequestWarning: Unverified HTTPS request is being made to host 's3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
2021-06-08 21:15:06 1024000000 small_chunk_1
2021-06-08 21:15:28 1024000000 small_chunk_1._read
2021-06-08 21:15:12 1024000000 small_chunk_10
2021-06-08 21:15:26 1024000000 small_chunk_10._read
2021-06-08 21:15:27 1024000000 small_chunk_1_read
2021-06-08 21:16:15 1024000000 small_chunk_2
2021-06-08 21:16:18 1024000000 small_chunk_2._read
2021-06-08 21:16:53 1024000000 small_chunk_3
2021-06-08 21:16:25 1024000000 small_chunk_3._read
2021-06-08 21:16:08 1024000000 small_chunk_4
2021-06-08 21:17:12 1024000000 small_chunk_4._read
2021-06-08 21:17:25 1024000000 small_chunk_5
2021-06-08 21:17:36 1024000000 small_chunk_5._read
2021-06-08 21:17:44 1024000000 small_chunk_6
2021-06-08 21:18:36 1024000000 small_chunk_6._read
2021-06-08 21:19:13 1024000000 small_chunk_7
2021-06-08 21:19:05 1024000000 small_chunk_7._read
2021-06-08 21:19:22 1024000000 small_chunk_8
2021-06-08 21:19:29 1024000000 small_chunk_8._read
2021-06-08 21:19:39 1024000000 small_chunk_9
2021-06-08 21:19:37 1024000000 small_chunk_9._read
[root@hpo-node1 akshat]# 
[root@hpo-node1 akshat]# 
[root@hpo-node1 akshat]# s3u1 ls s3://bucket-21/
urllib3/connectionpool.py:1013: InsecureRequestWarning: Unverified HTTPS request is being made to host 's3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
2021-06-08 21:30:55 1024000000 small_chunk_8
2021-06-08 21:30:55 1024000000 small_chunk_8._read
2021-06-08 21:30:55 1024000000 small_chunk_9
2021-06-08 21:30:55 1024000000 small_chunk_9._read
```

Command used to copy files:
`s3u1 cp --recursive s3://bucket-1 s3://bucket-21/`

Errors:
```
copy failed: s3://bucket-1/small_chunk_9 to s3://bucket-21/small_chunk_9 An error occurred (NoSuchKey) when calling the CompleteMultipartUpload operation (reached max retries: 2): The specified key does not exist.
copy failed: s3://bucket-1/small_chunk_8 to s3://bucket-21/small_chunk_8 An error occurred (NoSuchKey) when calling the CompleteMultipartUpload operation (reached max retries: 2): The specified key does not exist.
copy failed: s3://bucket-1/small_chunk_9._read to s3://bucket-21/small_chunk_9._read An error occurred (NoSuchKey) when calling the CompleteMultipartUpload operation (reached max retries: 2): The specified key does not exist.
```

### More information - Screenshots / Logs / Other output

```
Jun-9 4:29:35.711 [Endpoint/9]   [LOG] CONSOLE:: update_endpoint_stats. namespace_stats = []
Jun-9 4:30:05.205 [Endpoint/9]   [LOG] CONSOLE:: NamespaceFS: abort_object_upload /nsfs/nsfs-nsr-1/bucket-21/.noobaa-nsfs_60c0406ba0157e00233de2dc/multipart-uploads/f989c13a-82c4-40e7-b1ba-71162a3
def85
Jun-9 4:30:05.897 [Endpoint/9] [ERROR] CONSOLE::  [Error: No such file or directory] { code: 'ENOENT' }
Jun-9 4:30:05.897 [Endpoint/9] [ERROR] CONSOLE::  [Error: No such file or directory] { code: 'ENOENT' }
Jun-9 4:30:05.995 [Endpoint/9] [ERROR] core.endpoint.s3.s3_rest:: S3 ERROR <?xml version="1.0" encoding="UTF-8"?><Error><Code>NoSuchKey</Code><Message>The specified key does not exist.</Message><R
esource>/bucket-21/small_chunk_4._read?uploadId=f989c13a-82c4-40e7-b1ba-71162a3def85</Resource><RequestId>kpoz3eu6-6rwwuk-1aal</RequestId></Error> POST /bucket-21/small_chunk_4._read?uploadId=f989
c13a-82c4-40e7-b1ba-71162a3def85 {"accept-encoding":"identity","user-agent":"aws-cli/2.1.29 Python/3.8.8 Linux/4.18.0-305.3.1.el8_4.x86_64 exe/x86_64.rhel.8 prompt/off command/s3.cp","x-amz-date":
"20210609T042933Z","x-amz-content-sha256":"cc82905fa5a8bc75908709678e6972231fe14ce147ac9ef3611ec94ac4575f9f","authorization":"AWS4-HMAC-SHA256 Credential=YrpJ6lYpmcU6u8ODljQG/20210609/us-east-1/s3
/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=fad4887a005b0cd9e9fe190aadbbc283719e3d4af452b41385d8c04b97cf428e","content-length":"11641","host":"s3-noobaa.apps.ocp-a
kshat-1.cp.fyre.ibm.com","x-forwarded-host":"s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com","x-forwarded-port":"443","x-forwarded-proto":"https","forwarded":"for=10.17.66.193;host=s3-noobaa.apps.ocp
-akshat-1.cp.fyre.ibm.com;proto=https","x-forwarded-for":"10.17.66.193"} Error: No such file or directory
Jun-9 4:30:05.996 [Endpoint/9] [ERROR] core.endpoint.s3.s3_rest:: S3 ERROR <?xml version="1.0" encoding="UTF-8"?><Error><Code>NoSuchKey</Code><Message>The specified key does not exist.</Message><R
esource>/bucket-21/small_chunk_4._read?uploadId=f989c13a-82c4-40e7-b1ba-71162a3def85</Resource><RequestId>kpoz2p77-313rnn-1076</RequestId></Error> POST /bucket-21/small_chunk_4._read?uploadId=f989
c13a-82c4-40e7-b1ba-71162a3def85 {"accept-encoding":"identity","user-agent":"aws-cli/2.1.29 Python/3.8.8 Linux/4.18.0-305.3.1.el8_4.x86_64 exe/x86_64.rhel.8 prompt/off command/s3.cp","x-amz-date":
"20210609T042901Z","x-amz-content-sha256":"cc82905fa5a8bc75908709678e6972231fe14ce147ac9ef3611ec94ac4575f9f","authorization":"AWS4-HMAC-SHA256 Credential=YrpJ6lYpmcU6u8ODljQG/20210609/us-east-1/s3
/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=886bd10e1ca6928f0e01cb53ed17f8675de6e6d42d8ab351484e2dbf3f25c192","content-length":"11641","host":"s3-noobaa.apps.ocp-a
kshat-1.cp.fyre.ibm.com","x-forwarded-host":"s3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com","x-forwarded-port":"443","x-forwarded-proto":"https","forwarded":"for=10.17.66.193;host=s3-noobaa.apps.ocp
-akshat-1.cp.fyre.ibm.com;proto=https","x-forwarded-for":"10.17.66.193"} Error: No such file or directory
Jun-9 4:30:12.901 [Endpoint/9]   [LOG] CONSOLE:: NamespaceFS: abort_object_upload /nsfs/nsfs-nsr-1/bucket-21/.noobaa-nsfs_60c0406ba0157e00233de2dc/multipart-uploads/c33a6269-c625-4d62-9d9c-1b34258
23bd3
Jun-9 4:30:13.298 [Endpoint/9] [ERROR] CONSOLE::  [Error: No such file or directory] { code: 'ENOENT' }
```

**Comment by akmithal** [comment]
[noobaa_diagnostics_1623213530.tar.gz](https://github.com/noobaa/noobaa-core/files/6621193/noobaa_diagnostics_1623213530.tar.gz)

**Comment by jeniawhite** [comment]
Seems like a duplicate for this https://github.com/noobaa/noobaa-core/issues/6560

**Comment by nimrod-becker** [comment]
@akmithal ?

**Comment by liranmauda** [comment]
Regarding the Copy:
In the description of the issue we see that the copy is from one bucket to another 

> 
> ### Actual behavior
> 1. Copy operation from one bucket to another failed
> 

We can see in the block below that we are checking if we consider this as a server side copy (according to the below parameter): 

https://github.com/noobaa/noobaa-core/blob/052040a8b826e567d0d60847f741ff2e4a9554d6/src/sdk/namespace_fs.js#L158-L163

If we are not considering this as a server side copy we will do plain upload.

**Comment by nimrod-becker** [comment]
@akmithal so we understand why we do a read/write (different buckets, we don't want to do links between different paths since it might be different datasets).

And the failure with the read/write is the same a the multipart upload issue

**Comment by akmithal** [comment]
Another observation is, copy operation to same bucket fails as well. 

```
copy failed: s3://bucket-41/medium_file_1 to s3://bucket-41/medium_file_1_cp An error occurred (NoSuchKey) when calling the CompleteMultipartUpload operation: The specified key does not exist.

[root@hpo-node1 ~]# s3u1 cp s3://bucket-41/medium_file_1 s3://bucket-41/medium_file_1_cp
```

**Comment by akmithal** [comment]
[noobaa_diagnostics_1623649561.tar.gz](https://github.com/noobaa/noobaa-core/files/6645865/noobaa_diagnostics_1623649561.tar.gz)

**Comment by akmithal** [comment]
I am not confident if it is the same defect as 6560 as 6560 is not reproducible with 1 GB files as is the case for this defect. I think it should be still investigated. If root cause is found same, would retry this operation once fix for 6560 comes.

**Comment by jeniawhite** [comment]
@liranmauda Can you please take a look at the copy operation between the same buckets?
@akmithal Regarding the multipart upload (it looks like the same issue as we have in other issues).

**Comment by nimrod-becker** [comment]
We have other issues for the multipart, need to see if linking failed at the FS Level for this one (or why it failed)

**Comment by liranmauda** [comment]
By the logs, the link is not failing. 
as we are missing some prints we can assume that either link worked or copy flow was never called.

In #6582 I have added some prints in higher debug
In #6573 We enabled higher logging via an env variable 
In https://github.com/noobaa/noobaa-operator/pull/653 We have added the ability to install noobaa with a higher debug level.

Can you please reproduce with a higher debug level by installing with:  `noobaa install ... --debug-level <number>`
The number can be 2.

Also if the cluster is already running you and has the latest version you can edit the crd and add `debugLevel: <number>`

for example:
```
spec:
  cleanupPolicy: {}
  coreResources:
    limits:
      cpu: 100m
      memory: 1G
    requests:
      cpu: 100m
      memory: 1G
  dbImage: centos/postgresql-12-centos7
  dbResources:
    limits:
      cpu: 100m
      memory: 500M
    requests:
      cpu: 100m
      memory: 500M
  dbType: postgres
  debugLevel: 5
```

Tomorrow's build should contain all the required changes.

**Comment by akmithal** [comment]
Hi @liranmauda , Since in this case, it has attempted copy operation, either code is not written to handle Linking for namespace buckets OR Linking has failed and it went towards traditional copy operation. What is the code expected to do ?

**Comment by liranmauda** [comment]
We are doing copy in the following cases:
 - `copyObject` (and not `uploadPartCopy`)
 - If we get a `copyObject` we check if we consider it to be server side copy.
    - if it is not server side copy  then we treat it as `putObject`
    - If it is server side copy then we checking if there is a link for the source name and destination.
      - If there is one we return the etag, and consider it done.
      - If there is no link between  the requested source name and requested  destination we will try to link.
        - If link fails we will do a copy by open the files stream and close. 

When we try to link and fail into the "copy stream" we are printing this: 
https://github.com/noobaa/noobaa-core/blob/15f6113fcb9863d81abdb4d5700532c5557d95e7/src/sdk/namespace_fs.js#L542 

And as the warning above was not found on the logs we can assume that either the `putObject` flow was triggered or the link passed, or there is an already link between the source and destination, or `uploadPartCopy` was called.

**Comment by akmithal** [comment]
@jeniawhite , Please share your findings on whether you were able to reproduce it locally. Thanks.

**Comment by akmithal** [comment]
Hi @liranmauda , Me and Jenia found that we can see upto L4 logs in endpoint while doing copy operation. Kindly enable it in the code as well for me to reproduce this issue with higher debug info. 

```
[root@ocp-akshat-1-inf tmp_1]# podn
NAME                                               READY   STATUS        RESTARTS   AGE
noobaa-core-0                                      1/1     Running       0          69s
noobaa-db-pg-0                                     1/1     Running       11         41h
noobaa-default-backing-store-noobaa-pod-074d3314   0/1     Terminating   0          17h
noobaa-endpoint-5ff798f4fc-wssgx                   1/1     Running       0          111s
noobaa-operator-6cd984c444-86mnl                   1/1     Running       0          41h
[root@ocp-akshat-1-inf tmp_1]# oc describe pod noobaa-endpoint-5ff798f4fc-wssgx      |grep "NOOBAA_LOG_LEVEL:"
      NOOBAA_LOG_LEVEL:            5
```

**Comment by liranmauda** [comment]
@akmithal setting NOOBAA_LOG_LEVEL to a number specifies the max log level.
If you set it to be 5 and the highest there is, is 4, then you are good. 
If you reproduced it with the system above, please send the logs.

**Comment by akmithal** [comment]
Hi, Here is the observation and logs:

```
[root@ocp-akshat-1-inf ~]# podn
NAME                                               READY   STATUS        RESTARTS   AGE
noobaa-core-0                                      1/1     Running       0          4h47m
noobaa-db-pg-0                                     1/1     Running       11         46h
noobaa-default-backing-store-noobaa-pod-074d3314   0/1     Terminating   0          22h
noobaa-endpoint-5ff798f4fc-wssgx                   1/1     Running       0          4h48m
noobaa-operator-6cd984c444-86mnl                   1/1     Running       0          46h
[root@ocp-akshat-1-inf ~]# 
[root@ocp-akshat-1-inf ~]# 
[root@ocp-akshat-1-inf ~]# oc rsh noobaa-endpoint-5ff798f4fc-wssgx 
sh-4.4# cd /nsfs/nsfs-nsr-1/bucket-test3/
sh-4.4# ls -lrt
total 1000000
-rw-r-----. 1 5000 800          0 Jun 22 10:26 test
-rw-r--r--. 1 5000 800 1024000000 Jun 22 15:01 medium_file_1
```


```
[root@hpo-node1 medium_files_read]# s3u1 cp  s3://bucket-test3/medium_file_1 s3://bucket-test3/medium_file_1_copy
urllib3/connectionpool.py:1013: InsecureRequestWarning: Unverified HTTPS request is being made to host 's3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
urllib3/connectionpool.py:1013: InsecureRequestWarning: Unverified HTTPS request is being made to host 's3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
urllib3/connectionpool.py:1013: InsecureRequestWarning: Unverified HTTPS request is being made to host 's3-noobaa.apps.ocp-akshat-1.cp.fyre.ibm.com'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
fatal error: An error occurred (500) when calling the HeadObject operation (reached max retries: 2): Internal Server Error
```

**Comment by akmithal** [comment]
[noobaa_diagnostics_1624374406.tar.gz](https://github.com/noobaa/noobaa-core/files/6695470/noobaa_diagnostics_1624374406.tar.gz)

**Comment by liranmauda** [comment]
Tested with @akmithal 
Tested a 1 MB file -> linking is happening.
Tested 10 MB file -> linking is not happening.

The aws cli sends `MetadataDirective: REPLACE` if the file is bigger than 8 MB and in this case we are expecting not to link.

**Comment by guymguym** [comment]
@liranmauda Can you find this condition in the AWS SDK code? It's hard to understand why the object size should matter to the replacing-MD - perhaps something else is affecting this SDK decision?

**Comment by liranmauda** [comment]
> @liranmauda Can you find this condition in the AWS SDK code? It's hard to understand why the object size should matter to the replacing-MD - perhaps something else is affecting this SDK decision?

@guymguym 
Following the doc: https://docs.aws.amazon.com/cli/latest/reference/s3/cp.html

If the object is copied over in parts, the source object's metadata will not be copied over, no matter the value for the metadata-directive

TL;DR
> --metadata-directive (string) Specifies whether the metadata is copied from the source object or replaced with metadata provided when copying S3 objects. Note that if the object is copied over in parts, the source object's metadata will not be copied over, no matter the value for --metadata-directive, and instead the desired metadata values must be specified as parameters on the command line. Valid values are COPY and REPLACE. If this parameter is not specified, COPY will be used by default. If REPLACE is used, the copied object will only have the metadata values that were specified by the CLI command. Note that if you are using any of the following parameters: --content-type, content-language, --content-encoding, --content-disposition, --cache-control, or --expires, you will need to specify --metadata-directive REPLACE for non-multipart copies if you want the copied objects to have the specified metadata values.

**Comment by guymguym** [comment]
@liranmauda But what causes `aws s3 cp` to decide to copy in parts??

**Comment by liranmauda** [comment]
@guymguym 
The file size causes aws s3 cp to decide to copy in parts

https://aws.amazon.com/premiumsupport/knowledge-center/s3-multipart-upload-cli/

TL;DR
> It's a best practice to use aws s3 commands (such as aws s3 cp) for multipart uploads and downloads, because these aws s3 commands automatically perform multipart uploading and downloading based on the file size.

**Comment by nimrod-becker** [comment]
Closing the issue as we don't seem to have something to fix in our code, we can open a new one with a more accurate reason if we need it
